### PR TITLE
[HMA][MultiPart] Count matches per privacy group

### PR DESCRIPTION
**Pull request by schatten**
### Multi-part diff. Review in this order
* [Diff 1/5] [terraform: Move datastore definition to its own module](https://github.com/facebook/ThreatExchange/pull/595/commits/f3718518ef84e5588164f0aa5b336c46a6677db7)
* [Diff 2/5] [ Add methods to increment count of matches by privacy group ](https://github.com/facebook/ThreatExchange/pull/595/commits/671a2e5b3115d8c82191f3b100e215f26c53cd28)
* ~[Diff 3/5] [ Add PrivacyGroups to PDQ Match Record ](https://github.com/facebook/ThreatExchange/pull/595/commits/f7232a18d25079d5e60e88019e2be85d0f446d65)~ Reverted!
* [Diff 4/5] [ Stream-processing lambda: listens to DDB INSERT and MODIFY and updates counts](https://github.com/facebook/ThreatExchange/pull/595/commits/7224d326e7c08f70986cc180300e0f92d17276f4)
* ~[Diff 5/5] [ terraform: connect stream processing lambda to ddb updates ](https://github.com/facebook/ThreatExchange/pull/595/commits/c9cca15d2eae4e0d115714356ff7a6ec7dccd4ae)~ Reverted!

* [Diff 9/9] [ Second queue on matches SNS topic to increment pg counters ](https://github.com/facebook/ThreatExchange/pull/595/commits/3faafb04c4b90a0c6d6b7b40f2695870b4eea401)

Summary
---------
Adds a stream processor. This processor listens to updates to dynamodb and where required updates the counters of matches by privacy group.

A subsequent PR will show this information in the UI.

Test Plan
---------
New tests were added, so 
```
$ python -m py.test 
$ python -m mypy hmalib
$ python -m black hmalib
```

A fresh dynamodb instance was created and the storm script used to push some images to it.

Because of how my HMA instance is setup, the image matched 3 separate privacy groups.

As can be seen in the screenshot, all of their counters were incremented.
Screen Shot 2021-05-18 at 16 03 21

# Update

After the new changes using an SQS queue were done, I ran the storm test script against a different image to get counts updated
Screen Shot 2021-05-21 at 12 05 27
.

### CODE REVIEW

File: hasher-matcher-actioner/terraform/main.tf
Comment (BarrettOlson) [inline hasher-matcher-actioner/terraform/main.tf:329]: super nit/aside: do we want to have a standard auto formatter for tf files, I've noticed some inconsistencies (beside the whitespace in this PR) that could mess with the blame...

File: hasher-matcher-actioner/hmalib/tests/test_pdq_models.py
Comment (BarrettOlson) [inline hasher-matcher-actioner/hmalib/tests/test_pdq_models.py:396]: [Diff 2/5] nit: this test doesn't strike me as pdq specific, add a TODO note to refactor?

File: hasher-matcher-actioner/hmalib/common/count_models.py
Comment (BarrettOlson) [inline hasher-matcher-actioner/hmalib/common/count_models.py:21]: [Diff 2/5] scan could be unnecessarily pricey if we store this in the primary datastore. If we know can look up the SKs we want before had we should probably use them even if we want all counts... 
At the very least we could add a filter based on SK's prefix...

File: hasher-matcher-actioner/hmalib/lambdas/pdq/pdq_matcher.py
Comment (BarrettOlson) [inline hasher-matcher-actioner/hmalib/lambdas/pdq/pdq_matcher.py:123]: [Diff 3/5] `PDQSignalMetadata` has this information and can be looked up (we can add a helper to the class to make it easier).

Is your concern about retaining this info at all or at the very moment of the match?
- If at all I believe it is covered. 
- If at the very moment of the match then sure (Though I don't love the likely storage overlap and the different sources of truth to pick between presenting on the detail page but we can handle it) 

The fact that a signal is in a PG is not context the match record itself needs to care about imo; my understanding is that this is something `PDQSignalMetadata` and the match message would handle.

[not sure] but even if it is a at the very moment concern the match message downstream might be a better place to record this.

File: hasher-matcher-actioner/hmalib/lambdas/stream/ddb_stream_processor.py
Comment (BarrettOlson) [inline hasher-matcher-actioner/hmalib/lambdas/stream/ddb_stream_processor.py:62]: [Diff 4/5] Also add TODO note where total counters are currently implemented stating that they are going to be moved (to avoid someone copying that pattern in the meantime)

File: hasher-matcher-actioner/terraform/stream-processor/main.tf
Comment (BarrettOlson) [inline hasher-matcher-actioner/terraform/stream-processor/main.tf:70]: [Diff 5/5] feels like an over Allow, are all these actions being used? if not can we wait to add them until they are?

**Review by BarrettOlson — CHANGES_REQUESTED** [review]
Looks good! RC-ing to check understanding of the change and talk about where we want to retain PG info...

Possibly a dumb question but worth documenting the answer to even so:
- We are effectively creating a lambda that is triggered when a change to the datastore is made yes?
- This lambda (depending on the event) then updates the datastore
- Won't this cause the lambda to be triggered again? (if not what prevents that/can we add additional protections to do so)
  - It shouldn't infinite loop because the second event trigger will be ignored i.e. not update again but it still feels wasteful

File: hasher-matcher-actioner/hmalib/common/count_models.py
Comment (schatten) [reply hasher-matcher-actioner/hmalib/common/count_models.py:21]: I do not think it is pricey if the PK is the same. It will go to the same partition so the DTUs utilized should be limited.

File: hasher-matcher-actioner/hmalib/lambdas/pdq/pdq_matcher.py
Comment (schatten) [reply hasher-matcher-actioner/hmalib/lambdas/pdq/pdq_matcher.py:123]: You are right. This information is already present in PDQSignalMetadata. I'll rework to change how the information is inferred.

File: hasher-matcher-actioner/hmalib/tests/test_pdq_models.py
Comment (schatten) [reply hasher-matcher-actioner/hmalib/tests/test_pdq_models.py:396]: https://github.com/facebook/ThreatExchange/issues/594#issue-894714795

Does this work? Haha. Went to file an issue. Saw you'd already done that. Resolved.

File: hasher-matcher-actioner/terraform/stream-processor/main.tf
Comment (schatten) [reply hasher-matcher-actioner/terraform/stream-processor/main.tf:70]: Evidently all of these are necessary to be able to read ddb streams.

Comment (schatten) [reply hasher-matcher-actioner/terraform/stream-processor/main.tf:70]: Nope. Copy-pasted. I'll limit it to the read only ones.

File: hasher-matcher-actioner/hmalib/lambdas/match_counter.py
Comment (BarrettOlson) [inline hasher-matcher-actioner/hmalib/lambdas/match_counter.py:32]: ignore: is this comment just going to be in every lambda we write haha?

File: hasher-matcher-actioner/hmalib/lambdas/match_counter.py
Comment (BarrettOlson) [inline hasher-matcher-actioner/hmalib/lambdas/match_counter.py:29]: ignore: yo dawg, I heard you liked lambda

File: hasher-matcher-actioner/terraform/counters/main.tf
Comment (BarrettOlson) [inline hasher-matcher-actioner/terraform/counters/main.tf:112]: do we need more than "dynamodb:UpdateItem" for ["${var.datastore.arn}*"]? I am not seeing the lambda do anything else?

File: hasher-matcher-actioner/terraform/counters/variables.tf
Comment (BarrettOlson) [inline hasher-matcher-actioner/terraform/counters/variables.tf:48]: Could be missing something but it doesn't look like this or anything to do with the "config_table" is used but all are passed in. Could be from a copy paste and I think it is likely we will want to add it eventually to enable configurations after tf apply but could you note it as currently unused  in a comment if you want to keep it in the meantime? 

(my [not blocking] preference is to reduce the amount of things we give to what is being used for now)

File: hasher-matcher-actioner/terraform/main.tf
Comment (BarrettOlson) [inline hasher-matcher-actioner/terraform/main.tf:127]: .

**Review by BarrettOlson — APPROVED** [review]
Nice! approving to unblock. It feels like there is some leftover code from copying that ideally would get cleaned up for comment acknowledged before merge, however my/the teams other tf work is enough of a glass house that I don't think it is at all blocking haha. 

Thanks!

**Pull request merged**
